# Bayesian Analysis of Species Networks

A tutorial on Bayesian phylogenetic analysis of species networks

## Summary

XXX


## Table of contents

* [Outline](#outline)
* [Dataset](#dataset)
* [Requirements](#requirements)
* [Bayesian analysis of species networks with BEAST2](#speciesnetworks)


<a name="outline"></a>
## Outline

In this tutorial I am going to present how XXX.


<a name="dataset"></a>
## Dataset

The dataset used in this tutorial is the set of 68 alignments generated in tutorial [Local Assembly](../local_assembly/README.md). These 68 alignments contain gene sequences of five species of the cichlid genus *Neolamprologus* of Lake Tanganyika: *Neolamprologus brichardi* ("neobri"), *Neolamprologus gracilis* ("neogra") *Neolamprologus marunguensis* ("neomar"), *Neolamprologus olivaceous* ("neooli"), and *Neolamprologus pulcher* ("neopul"). The generation of these alignments is described in detail in tutorials [Ortholog Detection](../ortholog_detection/README.md) and [Local Assembly](../local_assembly/README.md). In brief, sequences of all species except *Neolamprologus pulcher* ("neopul") were extracted from whole-genome assemblies generated by [Brawand et al. (2014)](https://www.nature.com/articles/nature13726) and [Gante et al. (2016)](https://onlinelibrary.wiley.com/doi/abs/10.1111/mec.13767) in tutorial [Ortholog Detection](../ortholog_detection/README.md). Orthologous sequences of the fifth species, *Neolamprologus pulcher* ("neopul"), were identified and assembled from genomic read data in tutorial [Local Assembly](../local_assembly/README.md), and were then added to the alignments for the other four species. The set of alignments was then filtered so that the remaining 68 alignments that are used in this tutorial all contain sequence information for all five species and are each at least 500 bp in length.

<a name="requirements"></a>
## Requirements

* **BEAST2:** If you followed other tutorials in this collection, you likely have the BEAST2 package, including BEAST2 itself and TreeAnnotator, installed already. If not, you can downloaded these from the BEAST2 website [https://www.beast2.org](https://www.beast2.org). As all these programs are written in Java, compilation is not required, and all programs should work on Mac OS X, Linux, and Windows. I recommend downloading the program versions that include the Java Runtime Environment, which may prevent conflicts with Java versions that may already be installed on your machine.<br>

* **Tracer:** Like BEAST2, you probably have Tracer installed already if you did other tutorials of this collection. If not, you can download Tracer for Mac OS X, Linux, or Windows from [http://tree.bio.ed.ac.uk/software/tracer/](http://tree.bio.ed.ac.uk/software/tracer/). The file with the extension `.dmg` is for Mac OS X, the one with the extension `.tgz` is for Linux, and the Windows version is the file ending in `.zip`.



<a name="speciesnetworks"></a>
## Bayesian analysis of species networks with BEAST2

XXX

* Download the compressed directory [`10.tgz`](data/10.tgz) containing the 68 gene-sequence alignments generated in tutorial [Local Assembly](../local_assembly/README.md), either by clicking on the link or by using the following command:

		wget 


* XXX	**Question 1:** XXX? [(see answer)](#q1)


<br><hr>

<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>

## Answers

<a name="q1"></a>

* **Question 1:** XXX