# Bayesian Phylogenetic Inference

A tutorial on Bayesian inference of time-calibrated phylogenies

## Summary

XXX

## Table of contents

* [Outline](#outline)
* [Dataset](#dataset)
* [Requirements](#requirements)
* [Bayesian phylogenetic inference with BEAST2](#beast2)


<a name="outline"></a>
## Outline

In this tutorial, I will demonstrate how time-calibrated phylogenies can be inferred with programs of the Bayesian software package [BEAST2](https://www.beast2.org) ([Bouckaert et al. 2014](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003537)). The settings for the analysis will be specified with the program BEAUti, the Bayesian analysis itself is going to be conducted with BEAST2, and a summary tree will be generated with the program TreeAnnotator. These three programs are part of the BEAST2 package. In addition, I will present the use of the program [Tracer](http://tree.bio.ed.ac.uk/software/tracer/) ([Rambaut et al. 2018](https://academic.oup.com/sysbio/advance-article/doi/10.1093/sysbio/syy032/4989127)) to assess stationarity of the Bayesian analysis.


<a name="dataset"></a>
## Dataset

The data used in this tutorial are the filtered versions of the alignments generated for 16s and rag1 sequences in tutorial [Multiple Sequence Alignment](../multiple_sequence_alignment/README.md). These alignments contain sequence data for 41 teleost fish species and are 486 and 1,368 bp long, respectively. More information on the origin of the dataset can be found in the [Multiple Sequence Alignment](../multiple_sequence_alignment/README.md) tutorial. The software BEAST2 requires input in Nexus format, therefore we will use the files [`16s_filtered.nex`](data/16s_filtered.nex) and [`rag1_filtered.nex`](data/rag1_filtered.nex).

<a name="requirements"></a>
## Requirements

* **BEAST2:** The BEAST2 package, including BEAUti, BEAST2 itself, TreeAnnotator, and other tools can be downloaded from the BEAST2 website [https://www.beast2.org](https://www.beast2.org). As all these programs are written in Java, compilation is not required, and all programs should work on Mac OS X, Linux, and Windows. I recommend downloading the program versions that include the Java Runtime Environment, which may prevent conflicts with Java versions that may already be installed on your machine.<br>

* **Tracer:** Just like BEAST2, Tracer is written in Java and should work on your system without problems. The program can be downloaded for Mac OS X, Linux, or Windows from [http://tree.bio.ed.ac.uk/software/tracer](http://tree.bio.ed.ac.uk/software/tracer).

<a name="beast2"></a>
## Bayesian phylogenetic inference with BEAST2

In this part of the tutorial, we will run a basic Bayesian phylogenetic analysis for the 16s and rag1 alignments, using the programs of the BEAST2 software package. In this analysis, we are going to assume that the 16s and rag1 genes share the same evolutionary history and that this history is also identical to the evolutionary history of the 41 teleost fish species from which the sequences were obtained. In other words, we are going to assume that the two "gene trees" are identical and that they also are identical to the "species tree".

* If you're not familiar yet with Bayesian analyses and Markov-Chain Monte Carlo methods in general, you might be overwhelmed at first by the complexities of this type of analyses. Thus, it might be worth noting the many resources made available by BEAST2 authors that provide a wealth of information, that, even if you don't need them right now, could prove to be useful at a later stage. Thus, you might want to take a moment to explore the [BEAST2 website](https://www.beast2.org) and quickly browse through the [glossary of terms related to BEAST2 analyses](https://www.beast2.org/glossary/index.html). Note that the BEAST2 website also provides a [wide range of tutorials and manuals](https://www.beast2.org/tutorials/index.html). In addition, you can find many further tutorials on the [Taming the BEAST](https://taming-the-beast.org) website, where you will also find information about the excellent [Taming-the-BEAST workshops](https://taming-the-beast.org/workshops/). Finally, if you have further questions regarding BEAST2, you could have a look if somebody else already asked those questions on the very active [user forum](https://groups.google.com/forum/#!forum/beast-users), or you could ask these questions there yourself.

* Open the program BEAUti from the BEAST2 package, and import both the filtered 16s and rag1 alignments. To do so, click "Import Alignment" from the "File" menu and select the two files [`16s_filtered.nex`](data/16s_filtered.nex) and [`rag1_filtered.nex`](data/rag1_filtered.nex). The BEAUti window should then look as shown in the screenshot below.<p align="center"><img src="img/beauti1.png" alt="BEAUti" width="700"></p>

* Note that the BEAUti interface has six different tabs, of which (at the top of the window), the first one named "Partitions" is currently selected. We will browse through the other tabs later, but first need to specify settings regarding the partitioning in the currently open tab. Click on the row for the rag1 alignment to select it. Then, click the "Split" button at the bottom of the BEAUti window. As suggested, split the alignment into partitions "1 + 2 + 3", which will divide the alignment by codon position, and thus implement the partitioning scheme suggest by PAUP\* in tutorial [Substitution Model Selection](../substitution_model_selection/README.md). Click "OK". You should now see four rows in the "Partitions" tab of BEAUti, one for 16s and three for rag1, and each of the rag1 alignments should have a length of 456 sites:<p align="center"><img src="img/beauti2.png" alt="BEAUti" width="700"></p>

* Select all four rows at the same time and click on "Link Trees" near the top of the BEAUti window. This will force BEAST2 to use the same phylogeny for all partitions, which is equivalent to concatenating the sequences as we did for maximum-likelihood analyses with RAxML in tutorial [Maximum-Likelihood Phylogenetic Inference](../ml_phylogeny_inference/README.md). As explained in that tutorial the practice of concatenation may lead to bias in the phylogenetic inference, particularly if young and rapidly-diverging groups of taxa are investigated. However, in this first introduction to phylogenetic inference with BEAST2, we will ignore this potential issue and assume that the true gene trees for 16s and rag1 are in fact identical to each other and to the species tree of the 41 teleost fish species.<br>After clicking on "Link Trees", you should notice that all cells in the column to the right, under the heading "Tree" show the same value, "16s_filtered...", as in the screenshot below. This indicates that they now all share the same tree.<p align="center"><img src="img/beauti3.png" alt="BEAUti" width="700"></p>

* With all four rows still selected, also click on "Link Clock Models". This means that the clock model that we will select for 16s will also apply to all partitions of the rag1 gene. With the relaxed clock model that we will select, it means that some branches are allowed to evolve faster than other branches (= to have higher substitution rates than others), but that this variation in rates is not inferred separately for each gene. Thus, branches that are inferred to have a high rate in the 16s sequences will also receive a high rate for each of the rag1 partitions. Vice versa, a branch that is inferred to evolve comparatively slowly is assumed to evolve slowly both in 16s and in rag1. However, this branch will still be allowed to have a higher absolute rate in one partition compared to another partition because the branch rates specified by the clock model (one rate per branch) will still be multiplied by a partition-specific rate multiplier (so that in total we then have four rates per branch: one for each partition). A good justification for this linking of clock models is that in nature, the speed of the molecular clock often depends on factors that are species-specific, such as metabolism and generation time ([Moorjani et al. 2016](http://www.pnas.org/content/113/38/10607.long)). This means that a species with a short generation time will be expected to have a comparatively slow rate not only in one gene but in all its genes.<br>After clicking on "Link Clock Models", the BEAUti window should look as shown in the screenshot below. Note that all cells in the column with heading "Clock Model" now have the value "16s_filtered...".<p align="center"><img src="img/beauti4.png" alt="BEAUti" width="700"></p>

* The settings in the "Partitions" tab are now complete. Skip the next tab called "Tip Dates". This tab could be used to specify the times at which samples were taken, which allows time-calibration of viral phylogenies. But compared to the time scales over which the 41 fish sequences have diverged, the small difference in sampling times (a few years) is completely negligible, thus we can safely ignore these. Thus, click on the "Site Model" tab next. In this tab we can specify the substitution models for all four partitions. Select the 16s partition in the panel at the left and click on the drop-down menu that currently says "JC69". Instead of the Jukes-Cantor model, use the GTR model as suggested by the model selection that we did with PAUP\* in tutorial [Substitution Model Selection](../substitution_model_selection/README.md). Also specify "4" in the field for the "Gamma Category Count" two lines above, to use a gamma model of rate variation with four rate categories. This was also supported by the selection of substitution models in PAUP\*. The BEAUti window should then look as shown in the screenshot below.<p align="center"><img src="img/beauti5.png" alt="BEAUti" width="700"></p>

* Still in the "Site Model" tab, select all three rag1 partitions in the panel at the left of the window. The main part of the window should then show the option "Clone from 16s_filtered" as in the screenshot below. Click "OK" to use the same site model as for 16s for all rag1 partitions.<p align="center"><img src="img/beauti6.png" alt="BEAUti" width="700"></p>

* Next, click on the "Clock Model" tab. From the drop-down menu that currently says "Strict Clock", choose "Relaxed Clock Log Normal" instead. This is the most commonly used relaxed clock model in which substitution rates of individual branches are drawn from a lognormal distribution. The mean and standard deviation of this lognormal distribution will be estimated as part of the MCMC search. The model is described in [Drummond et al. (2006)](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.0040088), there it is named the "UCLN" model. Ignore the additional options ("Number of Discrete Rates" etc.) that appear after selecting the model.<p align="center"><img src="img/beauti7.png" alt="BEAUti" width="700"></p>

* Click on the "Priors" tab. From the drop-down menu at the very top of the window, select "Birth Death Model" instead of "Yule Model". By doing so we add a parameter to the model for the extinction rate. If we would choose the alternative Yule model ([Yule 1925](http://rstb.royalsocietypublishing.org/content/213/402-410/21?intcmp=trendmd)), we would assume that no extinction ever occurred in the history of teleost fishes. As this seems rather unrealistic, the birth-death model ([Gernhard 2008](https://www.sciencedirect.com/science/article/pii/S0022519308001811)) is in most cases the more appropriate choice. Nevertheless, results are in practice rather little affected by the choice of this prior.<p align="center"><img src="img/beauti8.png" alt="BEAUti" width="700"></p>

* Most of the other items shown in the "Prior" panel correspond to prior distributions placed on the parameters of the substitution models for the four partitions. You may keep the default priors for each of these parameters. However, to allow time calibration of the phylogeny, prior distributions for selected divergence times still need to be specified, otherwise BEAST2 would have very little information (only from the priors on speciation and mutation rates) to estimate branch lengths according to an absolute time scale. But before prior distributions can be placed on the divergence of certain clades, these clades must first be defined. This can be done at the bottom of the "Priors" tab. Thus, scroll down to the end of the list until you see the "+ Add Prior" button, as shown in the below screenshot.<p align="center"><img src="img/beauti9.png" alt="BEAUti" width="700"></p>

* Click on the "+ Add Prior" button to open the "Taxon set editor" pop-up window. Select all taxa from the list on the left of that window, and click the double-right-arrow symbol (`>>`) to shift them to the right side of the window. Then, click on the ID for zebrafish ("Danioxxrerioxx") to shift it back to the left with the double-left-arrow symbol (`<<`). This way, the ingroup, including all taxa except zebrafish is defined as a clade, so that the divergence of this clade can later be used for time calibration. The clade currently selected corresponds to the taxonomic group of "Euteleosteomorpha" ([Betancur-R. et al. 2017](https://bmcevolbiol.biomedcentral.com/articles/10.1186/s12862-017-0958-3)), thus, enter this name at the top of the pop-up window as the "Taxon set label", as shown in the below screenshot.<p align="center"><img src="img/beauti10.png" alt="BEAUti" width="700"></p>

* Repeat the last step for seven more well-defined taxonomic clades that are represented in our dataset:

	* "Neoteleostei": Select all taxa, then exclude "Danioxxrerioxx" and "Oncorhyketaxxx" again.
	* "Acanthopterygii": Select all taxa, then exclude "Ateleopjaponic", "Chloropagassiz", "Danioxxrerioxx", "Muraenomarmora", "Oncorhyketaxxx", "Percopstransmo", "Polymixjaponic", and "Zeusxxxfaberxx".
	* "Eupercaria": Include "Acanthapomotis", "Acropomjaponic", "Gasteroaculeat", "Niphonxspinosu", "Percalanovemac", "Percichtruchax", "Tautogoadspers", "Triacananomalu", and "Zancluscornutu".
	* "Perciformes": Include "Gasteroaculeat" and "Niphonxspinosu".
	* "Ovalentaria": Include "Ambassispcxxxx", "Aplochepanchax", "Cichlaxtemensi", "Ectodusdescamp", "Etroplumaculat", "Geophagbrasili", "Herichtcyanogu", "Maylandzebraxx", "Monocirpolyaca", "Mugilxxcephalu", "Neolampbrichar", "Opistogaurifro", "Oreochrnilotic", "Oryziaslatipes", "Paratilpolleni", "Paretromaculat", "Ptychocgrandid", "Pundaminyerere", and "Tylochrpolylep".
	* "Pseudocrenilabrinae" (African cichlids): Include "Ectodusdescamp", "Maylandzebraxx", "Neolampbrichar", "Oreochrnilotic", "Pundaminyerere", and "Tylochrpolylep".
	* "Cichlinae" (Neotropical cichlids): Include "Cichlaxtemensi", "Geophagbrasili", and "Herichtcyanogu".
	
	Specify that all of these clades should be monophyletic, using the checkbox at the right of the row corresponding to each clade. The bottom part of the list in the "Prior" tab should then look as in the below screenshot.<p align="center"><img src="img/beauti11.png" alt="BEAUti" width="700"></p>
	
* We still need to specify age constraints for each clade. Start with "Euteleosteomorpha": We know that the oldest fossil assigned to the clade "Euteleosteomorpha" is *Leptolepides sprattiformis*, and that this fossil is at least 152.1 million years old. This means that the clade "Euteleosteomorpha" is also at least 152.1 million years old. So we have a hard minimum boundary for its age, but the maximum boundary for its age is more difficult to identify. Given that the oldest fossil is at least 152.1 million years old it might be more probable that the clade originated 155 million years ago than, say, 900 million years ago. But even the latter age can not be excluded with certainty based on the 152.1 million-year-old fossil. Nevertheless, we can assign a prior probability distribution that decreases continously for older ages, to express our intuition that an age of 200 million years is more probable than an age of 300 million years, and that 300 million years is more probable than 400 million years etc. The lognormal prior distributions are well suited for this. How exactly the shape of this distribution should be chosen is a matter of controversy in the literature and no general solution has been found. For those interested, the introduction of my paper [Matschiner et al. (2017)](https://academic.oup.com/sysbio/article/66/1/3/2418030) describes this controversy (and the rest of the paper proposes a certain type of priors, implemented in the [CladeAge package for BEAST2](http://evoinformatics.eu/cladeage.htm), as a solution). For now, let’s use a lognormal distribution with a mean of 20 and a standard deviation of 1 for all clades. So, to specify this prior probability distribution for the age of "Euteleosteomorpha", click on the drop-down menu to the right of the button for "Euteleosteomorpha.prior" that currently says "[none]". Select "Log Normal" as shown in the next screenshot.<p align="center"><img src="img/beauti12.png" alt="BEAUti" width="700"></p>Then, click on the black triangle to the left of the button for "Euteleosteomorpha.prior". Specify "20.0" as the value for "M" (that is the mean of the prior distribution) and "1.0" as the value for "S" (that is the standard deviation). Importantly, set a tick in the checkbox for "Mean in Real Space", otherwise the specified value for the mean will be considered to be in log space (meaning that its exponent would be used). Next, specify the minimum age of the oldest fossil of "Euteleosteomorpha", "152.1", as the “offset”. In the plot to the right, you should then see that the distribution is centered around 160-170, with a hard minimum boundary at 152.1 and a "soft" maximum boundary (a tail of decreasing probability). Finally, activate the checkbox for "Use Originate" a bit further below to specify that the divergence time of this clade from this sister clade should be constrained, not the time when the lineages within this clade began to diverge. The BEAUti window should then look as shown in the below screenshot.<p align="center"><img src="img/beauti13.png" alt="BEAUti" width="700"></p>

* Repeat the above procedure for all other clades, using the same mean (20.0; don't forget to tick "Mean in Real Space" each time) and standard deviation (1.0), and the following minimum fossil ages as the offset:
	* "Neoteleostei": 125.0
	* "Acanthopterygii": 98.0
	* "Eupercaria": 83.6
	* "Perciformes": 50.8
	* "Ovalentaria": 49.1
	* "Pseudocrenilabrinae": 45.0
	* "Cichlinae": 39.9

	Also, don't forget to activate the checkbox for "Use Originate" each time. In the end, the BEAUti window should look as in the screenshot below.<p align="center"><img src="img/beauti14.png" alt="BEAUti" width="700"></p>
	
* Continue to the "MCMC" tab. Keep 10,000,000 as the chain length, but change the names of the output files again: Click on the triangle to the left of "tracelog" and specify "combined.log". In the next field for "Log Every", set the number to "10000" (instead of the default 1,000) so that only every 10,000th MCMC state is written to the file. Click on the triangle again, then click on the black triangle to the left of "treelog". Specify "combined.trees"
as the file name and again use "10000" as the number in the field for "Log Every". When the window looks as in the below screenshot, click on "Save" in BEAUti's "File" menu, and name the resulting file in XML format `combined.xml`.<p align="center"><img src="img/beauti15.png" alt="BEAUti" width="700"></p>

* Now, open the program BEAST2 and select the file [`combined.xml`](res/combined.xml) as input file, as shown in the screenshot below. When you click the "Run" button, BEAST2 will start the analysis, which should take around half an hour to a full hour.<p align="center"><img src="img/beast1.png" alt="BEAUti" width="500"></p>